### PR TITLE
Update jax to 0.6.2

### PR DESCRIPTION
**Pull request by paul0403**
**Context:**
Updating jax from 0.6.0 to 0.6.2. 

**Description of the Change:**
1. A few jaxpr internals started requiring a `source_info` argument. https://github.com/jax-ml/jax/pull/28103
   At any point in the program, the current source info at that point can be retrieved by `jax._src.source_info.current()`.
   The impacted functions are 
   - `DynamicJaxprTrace.to_jaxpr_tracer, new_arg`
   - Primitives with custom staging rules (https://github.com/jax-ml/jax/pull/29347). For us, this is `sample_p, counts_p, probs_p, state_p`. For core pennylane this is updated in https://github.com/PennyLaneAI/pennylane/pull/7887
2. Literals are now inlined into jaxpr equations https://github.com/jax-ml/jax/pull/28233. This has the following impacts:
   - Firstly, when printing jaxpr, the types of literals are now printed. This required a few lit-ish tests to update.
   - When topoligically sorting equations, the invars to equations are not necessaily all `Var`s, but can now be `Literal`s as well. In the sorting algorithm we tap into the internals of `Var`, but this would break if an invar is a `Literal`. This is actually very easy to fix, as literals do not effect topological ordering and do not need to be dealt with, so we just skip over them during sorting
3. In the same above jax commit, during tracing they now drop unused variables. Concretely, right before exiting `DynamicJaxprTrace.to_jaxpr (and to_jaxpr2)`, they now call `_drop_unused_var()` on the traced jaxpr. This is usually fine, but in our control flow's tracing, we trace the classical part of the inner region first before tracing the quantum tape (see [`HybridOp`](https://github.com/PennyLaneAI/catalyst/blob/main/frontend/catalyst/jax_tracer.py#L416)). This means the classical variables used by the quantum tape, which is "unused" at the end of the classical tracing, will be dropped, and breaks quantum tracing. So we patch `_drop_unused_var()` to do nothing when tracing the classical part of control flow ops.
4. `PytreeDef.make_from_node_data_and_children()` has been removed (with no replacement) https://github.com/jax-ml/jax/pull/29491. I do this manually now.

[sc-95175]

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -118,6 +118,12 @@ frontend:
 	$(PYTHON) -m pip uninstall -y pennylane
 	$(PYTHON) -m pip install -e . --extra-index-url https://test.pypi.org/simple $(PIP_VERBOSE_FLAG)
 	rm -r frontend/pennylane_catalyst.egg-info
+	# View jax commits at
```
Comment (paul0403) [inline Makefile]: changelog

File: Makefile [hunk 2]
```diff
@@ -118,6 +118,22 @@ frontend:
...
+	# 1d652ab7f4346a974cf5888ab23713e1a9c2ddba (Apr.17): the "source_info change"
+	# 492cd3d9313cfd45e8bd63a8f51aa63d92924cd5 (Apr.18): reverts the source_info change (1d652ab7)
+	# f1803bef692b6d91815b02ec643980d9ab9b5132 (Apr.22): Checkpoint, last commit of Apr.22
+	# d4dd0c4985cd8046a2b97efbd9798e0ebf397c8f (Apr.23): Checkpoint, last commit of Apr.23
+	# Somewhere here: reverts the "reverts the source_info change (492cd3d9)"
+	#   i.e. source_info is back in
+	# 351c13e5b6dfec90863ef64f137a28da7426f38a (Apr.26): Checkpoint, last commit of Apr.26
+	# ef015a0d3f8effa3d313c7d692d9241a54a99922 (Apr.29): the "inline literals" change
+	#pip install git+https://github.com/jax-ml/jax.git@603f73016cdd6c576b819524ea270972439ff5c5
+	pip install git+https://github.com/PennyLaneAI/pennylane.git@bump-jax-0.6.2
```
Comment (paul0403) [inline Makefile]: Clean this up after PL with 0.6.2 is on testpypi.

File: Makefile [hunk 3]
```diff
@@ -118,6 +118,22 @@ frontend:
...
+	# 1d652ab7f4346a974cf5888ab23713e1a9c2ddba (Apr.17): the "source_info change"
+	# 492cd3d9313cfd45e8bd63a8f51aa63d92924cd5 (Apr.18): reverts the source_info change (1d652ab7)
+	# f1803bef692b6d91815b02ec643980d9ab9b5132 (Apr.22): Checkpoint, last commit of Apr.22
+	# d4dd0c4985cd8046a2b97efbd9798e0ebf397c8f (Apr.23): Checkpoint, last commit of Apr.23
+	# Somewhere here: reverts the "reverts the source_info change (492cd3d9)"
+	#   i.e. source_info is back in
+	# 351c13e5b6dfec90863ef64f137a28da7426f38a (Apr.26): Checkpoint, last commit of Apr.26
+	# ef015a0d3f8effa3d313c7d692d9241a54a99922 (Apr.29): the "inline literals" change
+	#pip install git+https://github.com/jax-ml/jax.git@603f73016cdd6c576b819524ea270972439ff5c5
+	pip install git+https://github.com/PennyLaneAI/pennylane.git@875f0693a19fd9217305745bdbf945224151588e
```
Comment (paul0403) [inline Makefile]: I will update and remove the "do not merge" label once PL on 0.6.2 publishes on testpypi tonight.

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 4]
```diff
@@ -16,15 +16,16 @@
 This module contains public API functions for quantum operators which are not
 included in PennyLane, or whose behaviour needs to be adapted for Catalyst.
 """
-
 import copy
```
Comment (dime10) [inline frontend/catalyst/api_extensions/quantum_operators.py]: reverting
```suggestion

import copy
```

File: frontend/catalyst/jax_extras/tracing.py [hunk 5]
```diff
@@ -245,7 +252,7 @@ def __lt__(self, other):
 def jaxpr_pad_consts(jaxprs: List[Jaxpr]) -> List[ClosedJaxpr]:
     """Align the constants of Jaxpr programs. Return the list of corresponding programs accepting
     the same constants."""
-    newvar = gensym("_")
+    newvar = gensym()
```
Comment (dime10) [inline frontend/catalyst/jax_extras/tracing.py:256]: What is this change about?

File: frontend/catalyst/jax_tracer.py [hunk 6]
```diff
@@ -1066,11 +1069,48 @@ def trace_quantum_measurements(
...
+
+                    # Essentially, we want to fill the children into the nodes of the out_tree
+                    # i.e. we want out_tree.children() to be meas_return_trees_children
+                    # e.g. if out_tree is PyTreeDef({0: *, 1: *}),
+                    # and meas_return_trees_children is [PyTreeDef((*, *)), PyTreeDef(*)],
+                    # we want to update out_tree to PyTreeDef({0: (*, *), 1: *})
+                    #
+                    # Jax used to have a great util called make_from_node_data_and_children
+                    # but they removed it in 0.6.2, so we have to be a bit manual here
+                    def _make_from_node_data_and_children(node_data, children):
```
Comment (dime10) [inline frontend/catalyst/jax_tracer.py]: Is there a reason to define this function everytime we run `trace_quantum_measurements`?  It's probably nicer, and more readable, to define it statically somewhere else. Maybe even outside this module in `jax_extras` or something.

File: frontend/catalyst/jax_extras/tracing.py (hunk 5 above)
Comment (paul0403) [reply frontend/catalyst/jax_extras/tracing.py:256]: Ah I forgot about this in the PR description. They removed `gensym`'s suffix, now `gensym` just aliases `Var` https://github.com/jax-ml/jax/pull/29273/files#diff-713d667c14b1dd075ec025dbd1cd8ea222e42c8b4c094ae05373d4d05b276dc5

File: frontend/catalyst/jax_tracer.py (hunk 6 above)
Comment (paul0403) [reply frontend/catalyst/jax_tracer.py]: https://github.com/PennyLaneAI/catalyst/pull/1897/commits/cbdcb9843cac69c6da79eb7dd00c80ef6912b72b

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 4 above)
Comment (paul0403) [reply frontend/catalyst/api_extensions/quantum_operators.py]: https://github.com/PennyLaneAI/catalyst/pull/1897/commits/eccfd6fd5cfc0691d49838db321ce8d24ed457dc

File: frontend/catalyst/jax_extras/tracing.py (hunk 5 above)
Comment (dime10) [reply frontend/catalyst/jax_extras/tracing.py:256]: Hmm, I have no idea what it does (or rather why) 😅 Was this responsible for generating the `_` in the test with `a, b_ ...` and `a_, b ...` ?

**Review by dime10 — APPROVED** [review]
Thanks @paul0403 !

File: frontend/catalyst/jax_extras/tracing.py (hunk 5 above)
Comment (paul0403) [reply frontend/catalyst/jax_extras/tracing.py:256]: Yes, the string argument to `gensym` is printed out as the suffix.

File: frontend/catalyst/jax_extras/tracing.py [hunk 7]
```diff
@@ -923,3 +931,42 @@ def bind(self, *args, **params):
...
+    """
+    A helper to create a PytreeDef from node data and children.
+    Jax used to have a great util called make_from_node_data_and_children,
+    but they removed it in 0.6.2, so we have to be a bit manual here.
+
+    Essentially, we want to fill the `children` into the nodes of the tree we return
+    i.e. we want `out_tree_def.children()` to be `children`
+    e.g. if `out_tree_def` is PyTreeDef({0: *, 1: *}),
+    and `children` is [PyTreeDef((*, *)), PyTreeDef(*)],
+    we want to return PyTreeDef({0: (*, *), 1: *})
```
Comment (erick-xanadu) [inline frontend/catalyst/jax_extras/tracing.py:946]: Would this need an acknowledgement of JAX's authors? I.e., mention that this function was obtained from JAX?

Comment (paul0403) [reply frontend/catalyst/jax_extras/tracing.py:946]: No, this function was not obtained from JAX. This was 100% me. 

Just fyi, JAX's original `make_from_node_data_and_children` was a cpp function exposed by nanobind. It also uses a completely different logic (it actually traverses into the trees at a very low tree-impl level). Here I'm not tapping into the tree at all, and I'm just creating mock python objects with the same tree structure by `tree_unflatten` for the children and connect the children to the node data. 

See point 4 of the PR description.
